### PR TITLE
Azure improvements

### PR DESCRIPTION
**Pull request by thovoll**
# Description
* Adopt official Azure crates.
* Added backend tests and e2e tests.

Sample logs for concurrent writer test (tracing code not part of this PR):
```
Feb 22 05:48:23.952  WARN deltalake::delta: adls2://thovollazurerustsdk/test-delta-table-1645537700/_delta_log/_commit_fb4cf94f-045a-4a6d-8123-b49e820bdd7c.json.tmp - attempt_number = 0
Feb 22 05:48:23.955  WARN deltalake::delta: adls2://thovollazurerustsdk/test-delta-table-1645537700/_delta_log/_commit_10572b5b-5e0d-45ba-abb2-a9d0e7eb79df.json.tmp - attempt_number = 0
Feb 22 05:48:23.960  WARN deltalake::delta: adls2://thovollazurerustsdk/test-delta-table-1645537700/_delta_log/_commit_d6a87333-cfec-45c8-b5d0-24909264d921.json.tmp - attempt_number = 0
Feb 22 05:48:23.988  WARN deltalake::delta: adls2://thovollazurerustsdk/test-delta-table-1645537700/_delta_log/_commit_6412440f-c524-483b-897a-ac31651a1eb8.json.tmp - attempt_number = 0
Feb 22 05:48:24.079 ERROR azure_core::policies::retry_policies::retry_policy: server returned error status which will not be retried: 409 Conflict
Feb 22 05:48:24.089 ERROR azure_core::policies::retry_policies::retry_policy: server returned error status which will not be retried: 409 Conflict
Feb 22 05:48:24.129 ERROR azure_core::policies::retry_policies::retry_policy: server returned error status which will not be retried: 409 Conflict
Feb 22 05:48:24.175  WARN deltalake::delta: adls2://thovollazurerustsdk/test-delta-table-1645537700/_delta_log/_commit_d6a87333-cfec-45c8-b5d0-24909264d921.json.tmp - attempt_number = 1
Feb 22 05:48:24.180  WARN deltalake::delta: adls2://thovollazurerustsdk/test-delta-table-1645537700/_delta_log/_commit_10572b5b-5e0d-45ba-abb2-a9d0e7eb79df.json.tmp - attempt_number = 1
Feb 22 05:48:24.267  WARN deltalake::delta: adls2://thovollazurerustsdk/test-delta-table-1645537700/_delta_log/_commit_6412440f-c524-483b-897a-ac31651a1eb8.json.tmp - attempt_number = 1
Feb 22 05:48:24.290 ERROR azure_core::policies::retry_policies::retry_policy: server returned error status which will not be retried: 409 Conflict
Feb 22 05:48:24.398  WARN deltalake::delta: adls2://thovollazurerustsdk/test-delta-table-1645537700/_delta_log/_commit_05a1f6d6-5d72-48a1-a869-156415d33abc.json.tmp - attempt_number = 0
Feb 22 05:48:24.401  WARN deltalake::delta: adls2://thovollazurerustsdk/test-delta-table-1645537700/_delta_log/_commit_10572b5b-5e0d-45ba-abb2-a9d0e7eb79df.json.tmp - attempt_number = 2
Feb 22 05:48:24.538 ERROR azure_core::policies::retry_policies::retry_policy: server returned error status which will not be retried: 409 Conflict
Feb 22 05:48:24.627  WARN deltalake::delta: adls2://thovollazurerustsdk/test-delta-table-1645537700/_delta_log/_commit_05a1f6d6-5d72-48a1-a869-156415d33abc.json.tmp - attempt_number = 1
Feb 22 05:48:24.743  WARN deltalake::delta: adls2://thovollazurerustsdk/test-delta-table-1645537700/_delta_log/_commit_ef2efff2-8a7e-43ee-846a-e2683ebb86f8.json.tmp - attempt_number = 0
Feb 22 05:48:24.782  WARN deltalake::delta: adls2://thovollazurerustsdk/test-delta-table-1645537700/_delta_log/_commit_ff512a83-ec7a-48f8-ad27-a707fec1c0d8.json.tmp - attempt_number = 0
Feb 22 05:48:24.832 ERROR azure_core::policies::retry_policies::retry_policy: server returned error status which will not be retried: 409 Conflict
Feb 22 05:48:24.886  WARN deltalake::delta: adls2://thovollazurerustsdk/test-delta-table-1645537700/_delta_log/_commit_e5000ad0-4a5e-40f9-a782-efcd7683c3ec.json.tmp - attempt_number = 0
Feb 22 05:48:25.118  WARN deltalake::delta: adls2://thovollazurerustsdk/test-delta-table-1645537700/_delta_log/_commit_d3826e7d-059c-4ff5-86c5-fb6740bc6df4.json.tmp - attempt_number = 0
Feb 22 05:48:25.128  WARN deltalake::delta: adls2://thovollazurerustsdk/test-delta-table-1645537700/_delta_log/_commit_fc8d3ae1-b8ee-4d41-b8aa-c06dd3346c94.json.tmp - attempt_number = 0
Feb 22 05:48:25.162  WARN deltalake::delta: adls2://thovollazurerustsdk/test-delta-table-1645537700/_delta_log/_commit_ff512a83-ec7a-48f8-ad27-a707fec1c0d8.json.tmp - attempt_number = 1
Feb 22 05:48:25.199 ERROR azure_core::policies::retry_policies::retry_policy: server returned error status which will not be retried: 409 Conflict
Feb 22 05:48:25.259  WARN deltalake::delta: adls2://thovollazurerustsdk/test-delta-table-1645537700/_delta_log/_commit_548919b1-c6d1-4243-b450-709ababf176d.json.tmp - attempt_number = 0
Feb 22 05:48:25.287 ERROR azure_core::policies::retry_policies::retry_policy: server returned error status which will not be retried: 409 Conflict
Feb 22 05:48:25.394  WARN deltalake::delta: adls2://thovollazurerustsdk/test-delta-table-1645537700/_delta_log/_commit_62aad968-84a1-4f83-983a-bcb637515a90.json.tmp - attempt_number = 0
Feb 22 05:48:25.397  WARN deltalake::delta: adls2://thovollazurerustsdk/test-delta-table-1645537700/_delta_log/_commit_d3826e7d-059c-4ff5-86c5-fb6740bc6df4.json.tmp - attempt_number = 1
Feb 22 05:48:25.427  WARN deltalake::delta: adls2://thovollazurerustsdk/test-delta-table-1645537700/_delta_log/_commit_ff512a83-ec7a-48f8-ad27-a707fec1c0d8.json.tmp - attempt_number = 2
Feb 22 05:48:25.599 ERROR azure_core::policies::retry_policies::retry_policy: server returned error status which will not be retried: 409 Conflict
Feb 22 05:48:25.714  WARN deltalake::delta: adls2://thovollazurerustsdk/test-delta-table-1645537700/_delta_log/_commit_62aad968-84a1-4f83-983a-bcb637515a90.json.tmp - attempt_number = 1
Feb 22 05:48:25.719  WARN deltalake::delta: adls2://thovollazurerustsdk/test-delta-table-1645537700/_delta_log/_commit_4e8dec18-9613-45e5-8a9e-09b39ce5547f.json.tmp - attempt_number = 0
Feb 22 05:48:25.826 ERROR azure_core::policies::retry_policies::retry_policy: server returned error status which will not be retried: 409 Conflict
Feb 22 05:48:25.850 ERROR azure_core::policies::retry_policies::retry_policy: server returned error status which will not be retried: 409 Conflict
Feb 22 05:48:25.933  WARN deltalake::delta: adls2://thovollazurerustsdk/test-delta-table-1645537700/_delta_log/_commit_4e8dec18-9613-45e5-8a9e-09b39ce5547f.json.tmp - attempt_number = 1
Feb 22 05:48:25.961  WARN deltalake::delta: adls2://thovollazurerustsdk/test-delta-table-1645537700/_delta_log/_commit_ff512a83-ec7a-48f8-ad27-a707fec1c0d8.json.tmp - attempt_number = 3
Feb 22 05:48:26.019 ERROR azure_core::policies::retry_policies::retry_policy: server returned error status which will not be retried: 409 Conflict
Feb 22 05:48:26.106  WARN deltalake::delta: adls2://thovollazurerustsdk/test-delta-table-1645537700/_delta_log/_commit_ff512a83-ec7a-48f8-ad27-a707fec1c0d8.json.tmp - attempt_number = 4
Feb 22 05:48:26.199  WARN deltalake::delta: adls2://thovollazurerustsdk/test-delta-table-1645537700/_delta_log/_commit_ce5b0a0b-0847-49ba-9a54-eb509c9a66ee.json.tmp - attempt_number = 0
Feb 22 05:48:26.636  WARN deltalake::delta: adls2://thovollazurerustsdk/test-delta-table-1645537700/_delta_log/_commit_d35b5a51-cea2-47aa-9d5c-3f45c51350ad.json.tmp - attempt_number = 0
```

### CODE REVIEW

**Comment by thovoll** [comment]
Currently, there's a failing test due to https://github.com/Azure/azure-sdk-for-rust/issues/651

**Comment by thovoll** [comment]
create_table_and_commit() passes now.

File: rust/tests/adls_gen2_backend_test.rs
Comment (fvaleye) [inline rust/tests/adls_gen2_backend_test.rs:7]: It could be a follow-up: having dedicated Azure storage options would be really helpful if we want to configure the storage from other Bindings.

S3 Storage options: https://github.com/delta-io/delta-rs/blob/2fa81c611e74235c13e5d27813d2e9d815d10ad9/rust/src/storage/s3/mod.rs#L144

S3 Storage options in action with Python binding: https://github.com/delta-io/delta-rs/blob/main/python/docs/source/usage.rst#loading-a-delta-table

File: rust/tests/adls_gen2_backend_test.rs
Comment (fvaleye) [inline rust/tests/adls_gen2_backend_test.rs:2]: Could you share with me why the tests couldn't be run on the CI?

File: rust/tests/adls_gen2_backend_test.rs
Comment (thovoll) [reply rust/tests/adls_gen2_backend_test.rs:7]: Yes, I filed an issue for this: #555

File: rust/tests/adls_gen2_backend_test.rs
Comment (thovoll) [reply rust/tests/adls_gen2_backend_test.rs:2]: Yes, please see here: #497

File: rust/Cargo.toml
Comment (houqp) [inline rust/Cargo.toml:33]: @thovoll is the `needed` comment still relevant? Looks like something we should look into to avoid dependency creep.

**Review by houqp — APPROVED** [review]
Very cool @thovoll !

File: rust/Cargo.toml
Comment (thovoll) [reply rust/Cargo.toml]: Removed.

**Review by fvaleye — APPROVED** [review]
Thank you for your PR @thovoll!

LGTM 👍 

My only remaining suggestion would be to make sure to warn the developer by displaying a message [explanining](https://github.com/delta-io/delta-rs/pull/556/files#diff-77b691ca61433efa283b26187ecc9070f69647a43d5d82d4c73c14133a0f0af3R4) to run the Azure ignore's tests when the azure feature is enabled with `cargo test`. The CI will not be our safeguard anymore but we could accept to have potential regressions on future releases.

I'm really interested in your opinion on this @houqp 😄

File: rust/src/storage/azure/mod.rs
Comment (fvaleye) [inline rust/src/storage/azure/mod.rs:220]: Don't hesitate to create a [follow-up issue ](https://github.com/delta-io/delta-rs/issues/new/choose) and add the link in the codebase to track the work to do.

**Comment by thovoll** [comment]
> Thank you for your PR @thovoll!
> 
> LGTM 👍
> 
> My only remaining suggestion would be to make sure to warn the developer by displaying a message [explanining](https://github.com/delta-io/delta-rs/pull/556/files#diff-77b691ca61433efa283b26187ecc9070f69647a43d5d82d4c73c14133a0f0af3R4) to run the Azure ignore's tests when the azure feature is enabled with `cargo test`. The CI will not be our safeguard anymore but we could accept to have potential regressions on future releases.
> 
> I'm really interested in your opinion on this @houqp 😄

Thanks for your reviews @fvaleye! The end-to-end Azure tests were always ignored in CI actually, as described in #497. I'm going to soon work on the record/playback solution discussed in the same issue #497. Then, we'll have at least some protection in CI.

File: rust/src/storage/azure/mod.rs
Comment (thovoll) [reply rust/src/storage/azure/mod.rs:220]: Done! #568

**Comment by fvaleye** [comment]
> > Thank you for your PR @thovoll!
> > LGTM 👍
> > My only remaining suggestion would be to make sure to warn the developer by displaying a message [explanining](https://github.com/delta-io/delta-rs/pull/556/files#diff-77b691ca61433efa283b26187ecc9070f69647a43d5d82d4c73c14133a0f0af3R4) to run the Azure ignore's tests when the azure feature is enabled with `cargo test`. The CI will not be our safeguard anymore but we could accept to have potential regressions on future releases.
> > I'm really interested in your opinion on this @houqp 😄
> 
> Thanks for your reviews @fvaleye! The end-to-end Azure tests were always ignored in CI actually, as described in #497. I'm going to soon work on the record/playback solution discussed in the same issue #497. Then, we'll have at least some protection in CI.

Thank you for sharing more details with me 👍 
I wait for @houqp approval before merging your PR 🚀

**Review by houqp — APPROVED** [review]
LGTM!

**Comment by houqp** [comment]
good news that we won't need to maintain a release branch anymore.

**Pull request merged**
